graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+            in the incoming edges DataFrame.
+
+        Returns
+        -------
+        DataFrame
+            Returns a new mapping DataFrame, ordered by the updated
+            integer node ids.
+        """
+        join_col = NODE_MAPPING_STR
+        index_col = NODE_MAPPING_INT
+        incoming_node_ids = (
+            edges_df.select(node_col).distinct().withColumnRenamed(node_col, join_col)
+        )
+
+        # TODO: When is it possible that we might get a null node str id value?
+        # Is it only in the case the actual row has null values?
+        node_df_with_ids = incoming_node_ids.join(mapping_df, on=join_col, how="fullouter")
+
+        # If we have new node ids we should re-index
+        # TODO: The counts trigger computations, should we avoid this?
+        if node_df_with_ids.count() > mapping_df.count():
+            logging.info(
+                "Node mapping count mismatch, for node_col: %s, re-assigning int ids...", node_col
+            )
+            node_rdd_with_ids = (
+                node_df_with_ids.select(join_col)
+                .rdd.zipWithIndex()
+                .map(lambda rdd_row: (list(rdd_row[0]) + [rdd_row[1]]))
+            )
+
+            map_schema = StructType(
+                [
+                    StructField(join_col, StringType(), True),
+                    StructField(index_col, IntegerType(), True),
+                ]
+            )
+
+            node_df_with_ids = self.spark.createDataFrame(node_rdd_with_ids, map_schema)
+
+        # TODO: Do not perform ordering here but only once at end?
+        # Would it actually make a difference or is this optimized away?
+        return node_df_with_ids.orderBy(NODE_MAPPING_INT)
+
+    def create_node_id_map_from_nodes_df(self, node_df: DataFrame, node_col: str) -> DataFrame:
+        """Given a nodes DF will attach sequential node ids, order the DF by them,
+        and return a new DF with int id attached.
+
+        The mapping DF will have two columns, `node_str_id`
+        containing the original id, and `node_int_id`
+        that has the corresponding index.
+
+        Parameters
+        ----------
+        node_df : DataFrame
+            DataFrame that contains all the data for a node type
+        node_col : str
+            The column in the DataFrame that corresponds to the string ID of the nodes
+
+        Returns
+        -------
+        DataFrame
+            The input nodes DataFrame, with additional integer ids attached for each row, ordered
+            by the integer ids.
+        """
+        original_schema = node_df.schema
+
+        assert node_col in node_df.columns, f"{node_col=} not in {node_df.columns}"
+
+        # This will trigger a computation
+        node_rdd_with_ids = node_df.rdd.zipWithIndex()
+
+        node_id_col = f"{node_col}-int_id"
+        new_schema = original_schema.add(StructField(node_id_col, IntegerType(), False))
+
+        node_rdd_with_ids = node_rdd_with_ids.map(
```

Comment:
For node files we can assume that there's only one file source per node type, and that it must include all node ids that appear.

For when creating node mappings from edges I agree we do it that way, but it would likely hurt scalability a bit (if we maintain all unique node string ids in cluster memory) and would need a bit of re-writing as mentioned in another comment. We can leave it as follow-up work.